### PR TITLE
fix: statuspage incidents appear in message banner

### LINKED ISSUE #1182: Maintenance banner does not appear for incidents
## Current behavior

The notification bar at the top of the website appears in a few different circumstances, including when scheduled maintenance is occurring. However, it does not appear when an incident is reported (this was [always the behavior](https://github.com/web3-storage/web3.storage/blob/302cf2f5f45a4c8499dafcd6bf3df8e77e386e90/packages/website/components/layout.js#L39).

This content comes from a statuspage endpoint, [status.web3.storage/api/v2/summary.json](https://status.web3.storage/api/v2/summary.json).

## Expected behavior

- If there's an incident that is not resolved, show its description body in the bar
  - A caution icon should precede the text
  - (Needs to be verified on mobile and for multiline text)
- If there are multiple incidents, show the most recent
- If there is scheduled maintenance and incident(s) simultaneously, the most recent incident supersedes

***

### Example status object with an incident

```json


{
  "page": {
    "id": "93qc8z4vnx93",
    "name": "Web3 Storage",
    "url": "https://status.web3.storage",
    "time_zone": "Etc/UTC",
    "updated_at": "2022-03-30T00:32:56.297Z"
  },
  "components": [
    {
      "id": "wzppx4zzx3x5",
      "name": "API",
      "status": "partial_outage",
      "created_at": "2021-07-21T12:04:31.011Z",
      "updated_at": "2022-03-30T00:06:37.789Z",
      "position": 3,
      "description": null,
      "showcase": true,
      "start_date": "2021-07-21",
      "group_id": null,
      "page_id": "93qc8z4vnx93",
      "group": false,
      "only_show_if_degraded": false
    },
    {
      "id": "fy2sj2v6876n",
      "name": "Website",
      "status": "operational",
      "created_at": "2021-07-21T12:04:42.338Z",
      "updated_at": "2022-02-24T17:55:28.313Z",
      "position": 4,
      "description": null,
      "showcase": true,
      "start_date": "2021-07-21",
      "group_id": null,
      "page_id": "93qc8z4vnx93",
      "group": false,
      "only_show_if_degraded": false
    }
  ],
  "incidents": [
    {
      "id": "km6g0w0fpjk1",
      "name": "Reads unreliable for content recently uploaded Web3.Storage",
      "status": "investigating",
      "created_at": "2022-03-30T00:06:37.748Z",
      "updated_at": "2022-03-30T00:06:37.847Z",
      "monitoring_at": null,
      "resolved_at": null,
      "impact": "major",
      "shortlink": "https://stspg.io/kq62n14xnbfv",
      "started_at": "2022-03-30T00:06:37.742Z",
      "page_id": "93qc8z4vnx93",
      "incident_updates": [
        {
          "id": "q3g0zc9frhqk",
          "status": "investigating",
          "body": "Currently, we are seeing an increased number of reports of users unable to read data they have just uploaded to Web3.Storage from the IPFS network, including from gateways. We are investigating the issue further.",
          "incident_id": "km6g0w0fpjk1",
          "created_at": "2022-03-30T00:06:37.843Z",
          "updated_at": "2022-03-30T00:06:37.843Z",
          "display_at": "2022-03-30T00:06:37.843Z",
          "affected_components": [
            {
              "code": "wzppx4zzx3x5",
              "name": "API",
              "old_status": "operational",
              "new_status": "partial_outage"
            }
          ],
          "deliver_notifications": true,
          "custom_tweet": null,
          "tweet_id": null
        }
      ],
      "components": [
        {
          "id": "wzppx4zzx3x5",
          "name": "API",
          "status": "partial_outage",
          "created_at": "2021-07-21T12:04:31.011Z",
          "updated_at": "2022-03-30T00:06:37.789Z",
          "position": 3,
          "description": null,
          "showcase": true,
          "start_date": "2021-07-21",
          "group_id": null,
          "page_id": "93qc8z4vnx93",
          "group": false,
          "only_show_if_degraded": false
        }
      ]
    }
  ],
  "scheduled_maintenances": [],
  "status": {
    "indicator": "minor",
    "description": "Minor Service Outage"
  }
}

```

(cc @dchoi27)

### PR DESCRIPTION
**Pull request by orvn**
Adds functionality where, if an incident exists on StatusPage, it appears in the message banner. Previously, only _scheduled_ maintenance would appear. Described further in the linked issue. closes #1182.

### CODE REVIEW

**Comment by orvn** [comment]
@dchoi27, I'm going to simulate a few different scenarios locally to make sure this passes all conditions (multiple incidents, coinciding with maintenance, etc.).

I added you as a reviewer, but it's kind of hard for you to see what'll happen because the status page endpoint is clear right now. I'll take some screenshots of what it would look like on desktop and mobile, and maybe you can review based on that?

**Comment by dchoi27** [comment]
SGTM - looping in @yusefnapora and @gobengo to help review the code too

File: packages/website/components/messagebanner/messagebanner.js [hunk 1]
```diff
@@ -24,16 +24,19 @@ export default function MessageBanner() {
...
       (/** @type {{ status: string; }} */ maintenance) => maintenance.status !== 'completed'
     ) || [];
-
   const { data: apiVersionData, error: apiVersionError } = useQuery('get-version', () => getVersion(), {
     enabled: (statusPageData && scheduledMaintenances.length === 0) || statusPageError !== null,
   });
 
-  if (scheduledMaintenances.length > 0) {
+  if (incidents.length > 0) {
+    maintenanceMessage = statusPageData?.incidents[0].incident_updates[0].body;
```
Comment (gobengo) [inline packages/website/components/messagebanner/messagebanner.js:38]: What if perhaps `statusPageData?.incidents[0].incident_updates` expression evaluates to an empty array? Then I worry that the following `[0].body` could lead to a '.body does not exist on undefined' error at runtime.

What do you think of adding more optional chaining operators:
`statusPageData?.incidents?.[0].incident_updates?.[0]?.body`

```
> statusPageData = { incidents: [{ incident_updates: [] }] }
{ incidents: [ { incident_updates: [] } ] }
> statusPageData?.incidents[0].incident_updates[0].body
Uncaught TypeError: Cannot read property 'body' of undefined
> statusPageData?.incidents?.[0].incident_updates?.[0]?.body
undefined
> statusPageData = { incidents: [{ incident_updates: [{ body: 'foo' }] }] }
{ incidents: [ { incident_updates: [Array] } ] }
> statusPageData?.incidents?.[0].incident_updates?.[0]?.body
'foo'
```

**Comment by orvn** [comment]
@dchoi27 actually I do have a question: right now I have it set to display the body of the incident, which could be long in some cases. Yesterday we had:

> _Currently, we are seeing an increased number of reports of users unable to read data they have just uploaded to Web3.Storage from the IPFS network, including from gateways. We are investigating the issue further._

This is the behavior for scheduled maintenance as well, and can look a little long on mobile:

mobile-incident-message
desktop-incident-message

Alternatively, we could use the incident type and link out to it? It would just say

`⚠ Minor Service Outage`

File: packages/website/components/messagebanner/messagebanner.js (hunk 1 above)
Comment (orvn) [reply packages/website/components/messagebanner/messagebanner.js:38]: From what I had seen, if the incident exists, it always has a body, even if that body is an empty string. 

> What do you think of adding more optional chaining operators

_But_ we never know what might happen to an external source, so this makes sense to me and will help handle some exceptions, thanks!

Comment (orvn) [reply packages/website/components/messagebanner/messagebanner.js:38]: (also will add it to maintenance messages, which use a similar schema but are another object)

**Comment by dchoi27** [comment]
i think the extended description version is fine!

File: packages/website/components/messagebanner/messagebanner.js (hunk 1 above)
Comment (gobengo) [reply packages/website/components/messagebanner/messagebanner.js:38]: nice ty.

also curious what you or others think about techniques like asserting the schema of the json response to get type safety. I think ajv could support thing like minimum array length assertions. https://ajv.js.org/

File: packages/website/components/messagebanner/messagebanner.js [hunk 2]
```diff
@@ -24,16 +24,19 @@ export default function MessageBanner() {
 
   let maintenanceMessage = '';
 
+  const incidents = statusPageData?.incidents || [];
+
   const scheduledMaintenances =
     statusPageData?.scheduled_maintenances.filter(
       (/** @type {{ status: string; }} */ maintenance) => maintenance.status !== 'completed'
     ) || [];
```
Comment (alanshaw) [inline packages/website/components/messagebanner/messagebanner.js:32]: Default to empty array _before_ using `filter`.

```suggestion
    (statusPageData?.scheduled_maintenances || []).filter(
      (/** @type {{ status: string; }} */ maintenance) => maintenance.status !== 'completed'
    );
```

File: packages/website/components/messagebanner/messagebanner.js (hunk 1 above)
Comment (alanshaw) [inline packages/website/components/messagebanner/messagebanner.js:38]: Shorter

```suggestion
    maintenanceMessage = incidents[0].incident_updates[0].body;
```

File: packages/website/components/messagebanner/messagebanner.js [hunk 3]
```diff
@@ -24,16 +24,19 @@ export default function MessageBanner() {
...
-
   const { data: apiVersionData, error: apiVersionError } = useQuery('get-version', () => getVersion(), {
     enabled: (statusPageData && scheduledMaintenances.length === 0) || statusPageError !== null,
   });
 
-  if (scheduledMaintenances.length > 0) {
+  if (incidents.length > 0) {
+    maintenanceMessage = statusPageData?.incidents[0].incident_updates[0].body;
+  } else if (scheduledMaintenances.length > 0) {
     maintenanceMessage = statusPageData.scheduled_maintenances[0].incident_updates[0].body;
```
Comment (alanshaw) [inline packages/website/components/messagebanner/messagebanner.js:40]: You might have filtered out a completed maintenance above and then access it here.

```suggestion
    maintenanceMessage = scheduledMaintenances[0].incident_updates[0].body;
```

**Review by alanshaw — CHANGES_REQUESTED** [review]

**Review by gobengo — APPROVED** [review]
lgtm!

**Review by alanshaw — APPROVED** [review]
LGTM - not a fix for this PR but it was unexpected (for me) to see the body content here. IMHO we should be showing the (short) title of the incident/maintenance and users should click through to learn more. It allows us to go into much more detail in the extended description without having to worry about how much vertical space we're using on the website.

**Comment by orvn** [comment]
> LGTM - not a fix for this PR but it was unexpected (for me) to see the body content here. IMHO we should be showing the (short) title of the incident/maintenance and users should click through to learn more. It allows us to go into much more detail in the extended description without having to worry about how much vertical space we're using on the website.

Merging for now and opening a new issue for using the message `name` instead of the `body`

**Pull request merged**